server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java

```diff
@@ -81,6 +83,33 @@
      * Returns the mapped field type for the given field name.
      */
     MappedFieldType get(String field) {
+        GuardedFieldTypeLookup guardedLookup = new GuardedFieldTypeLookup();
+        return guardedLookup.get(field);
+    }
+
+    private class GuardedFieldTypeLookup {
+
+        Set<String> fieldPath = new LinkedHashSet<>();
+
+        MappedFieldType get(String field) {
+            if (fieldPath.contains(field)) {
+                throw new IllegalStateException("Loop in field resolution detected: " + String.join("->", fieldPath) + "->" + field);
+            }
+            fieldPath.add(field);
+            RuntimeFieldType runtimeFieldType = FieldTypeLookup.this.runtimeFields.get(field);
+            if (runtimeFieldType != null) {
+                return runtimeFieldType.asMappedFieldType(this::get);
+            }
+            return getConcrete(field);
+        }
+    }
+
+    /**
+     * Returns the mapped field type for the given field, excluding runtime fields
+     * @param field the field name
+     * @return a mapped field type, or null if the field does not exist
+     */
+    MappedFieldType getConcrete(String field) {
```

Comment:
I am a bit on the fence on these changes. I really liked that there is no differentiation between indexed and runtime fields when it comes to MappedFieldType. We already have FieldMapper on the other side that points to a mapped field type, and although I see why aliases may need this additional step, it was nice that it was not needed up until now, especially as MappedFieldType will become SearchField and represent the search behaviour of a field.

Reply:
> no differentiation between indexed and runtime fields when it comes to MappedFieldType

But for this specific usecase (using MappedFieldType to define an index sort) then we *do* need the distinction.  Though maybe `buildIndexSort` needs to use a FieldMapper rather than a MappedFieldType? Then we could remove this method.